src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+        long rowTombstones = Math.min(tombstoneCounts.rowTombstones, totalDataRows);
+
+        // we are over eastimating the garbage. so assuming that range tombstones are for the remaining undeleted partitions.
+        long rangeTombstones = Math.min(tombstoneCounts.rangeTombstones / 2, totalDataPartitions - partitionTombstones);
+        return (partitionTombstones * avgPartitionSize) + (rowTombstones * avgRowSize) + (rangeTombstones * avgPartitionSize / 3.0);
+    }
+
+    static double getEstimatedGarbage(OverallStats os)
+    {
+        return getEstimatedGarbage(os.tombstoneCounts, os.keyCount, os.rowCount, os.onDiskLength);
+    }
+
+    private synchronized SSTableStats getOrCompute(SSTableReader sstable)
+    {
+        if (!sstables.containsKey(sstable))
+        {
+            SSTableStats stats = new SSTableStats(sstable);
+            sstables.put(sstable, stats);
+            return stats;
+        }
+        return sstables.get(sstable);
+    }
+
+    @Override
+    public synchronized void addSSTable(SSTableReader sstable)
+    {
+        getOrCompute(sstable);
+    }
+
+    @Override
+    public synchronized void removeSSTable(SSTableReader sstable)
+    {
+        sstables.remove(sstable);
+    }
+
+    public static TimeWindow getTimeWindowMs(SSTableReader ssTableReader, long windowSizeInMs)
+    {
+        StatsMetadata metadata = ssTableReader.getSSTableMetadata();
+        return toWindowMs(metadata.minKey, metadata.maxKey, windowSizeInMs);
+    }
+
+    static TimeWindow toWindowMs(long tsStartInclusiveInMs, long tsEndExlusiveInMs, long windowSizeInMs)
+    {
+        long lowerWindowBound = toWindowMs(tsStartInclusiveInMs, windowSizeInMs);
+        long upperWindowBound = toWindowMs(tsEndExlusiveInMs, windowSizeInMs) + (tsEndExlusiveInMs % windowSizeInMs == 0 ? 0 : windowSizeInMs);
+
+        return new TimeWindow(lowerWindowBound, upperWindowBound - lowerWindowBound);
+    }
+
+    static long toWindowMs(long timeInMs, long windowSizeMs)
+    {
+        return (timeInMs / windowSizeMs) * windowSizeMs;
+    }
+
+    private void updateEstimatedCompactionsByTasks(Iterable<SSTableReader> sstables)
+    {
+        long windowSizeMs = getWindowSizeInMs();
+
+        Set<Long> tWindowSet = new HashSet<>();
+        Set<Long> dWindowSet = new HashSet<>();
+        int tWindowOccupancy = 0, dWindowOccupancy = 0;
+
+        for (SSTableReader sstable : sstables)
+        {
+            TimeWindow tw = getTimeWindowMs(sstable, windowSizeMs);
+            if (sstable.getSSTableLevel() == SSTable.DATA_SSTABLE_LVL)
+            {
+                dWindowOccupancy += collectOccupacyStat(dWindowSet, windowSizeMs, tw);
+            }
+            else
+            {
+                tWindowOccupancy += collectOccupacyStat(tWindowSet, windowSizeMs, tw);
+            }
+        }
+
+        int n = tWindowOccupancy - (2 * tWindowSet.size()) + dWindowOccupancy - (2 * dWindowSet.size());
```

Comment:
Can you explain this?

Reply:
even i dont know this :p 
This method needs to be reworked and i am looking for some clever way to calculate the pending tasks.